Uplift `structural_traits.rs` into the new trait solver

compiler-errors opened the pull request:
Self-explanatory. I will leave some comments inline regarding design decisions.

compiler-errors committed:
Rework most of structural_traits to be Interner-agnostic

compiler-errors committed:
Finish uplifting all of structural_traits

compiler-errors committed:
LangItem-ify Coroutine trait in solvers

compiler-errors committed:
Fix some TODOs

compiler-errors commented on compiler/rustc_infer/src/infer/at.rs:
```diff
@@ -225,6 +225,50 @@ impl<'a, 'tcx> At<'a, 'tcx> {
+    {
+        let mut fields = CombineFields::new(
+            self.infcx,
+            TypeTrace::dummy(self.cause),
```
Eventually I can uplift `TypeRelating` and we can get rid of this.

compiler-errors commented on compiler/rustc_type_ir/src/lang_items.rs:
```diff
@@ -0,0 +1,8 @@
+/// Lang items used by the new trait solver. This can be mapped to whatever internal
+/// representation of `LangItem`s used in the underlying compiler implementation.
+pub enum TraitSolverLangItem {
```
This is what I'm doing for now to deal with `LangItem`s. I'm generally okay with it since we're only using a subset of lang items in the solver.

compiler-errors commented on compiler/rustc_type_ir/src/interner.rs:
```diff
@@ -153,6 +148,38 @@ pub trait Interner:
+    fn item_bounds(
+        self,
+        def_id: Self::DefId,
+    ) -> ty::EarlyBinder<Self, impl Iterator<Item = Self::Clause>>;
```
For queries that just return lists of things, I'm not using an associated type. I'm just using an RPITIT -- this seems to me much easier than defining an associated type.

I'll keep associated types to things that we expect should have inherent methods or need to be named for other reasons.

compiler-errors commented on compiler/rustc_type_ir/src/interner.rs:
```diff
@@ -153,6 +148,38 @@ pub trait Interner:
+
+    fn require_lang_item(self, lang_item: TraitSolverLangItem) -> Self::DefId;
+
+    fn associated_type_def_ids(self, def_id: Self::DefId) -> impl Iterator<Item = Self::DefId>;
```
Specifically, an RPITIT is useful for a query like this because all we want is associated *type* def ids, and in the implementation we use `associated_item_def_ids` and `filter`.

compiler-errors commented on compiler/rustc_type_ir/src/infcx.rs:
```diff
@@ -1,23 +1,75 @@
+
+    fn relate<T: Relate<Self::Interner>>(
+        &self,
+        param_env: <Self::Interner as Interner>::ParamEnv,
```
unrelated but it's so annoying to have to refer to the interner like `<Self::Interner as Interner>::ParamEnv`. I wonder why we don't allow chaining associated type lookup like that...

fmease replied:
I guess it should be relatively straightforward to extend `<dyn HirTyLowerer>::lower_assoc_path` to look at the `item_bounds` of the inner assoc ty. Not sure if T-type's blessing would suffice or if it would require T-lang's. CC #22519 (arbitrary shorthand projections).

compiler-errors replied:
I think it would be awesome if we did that.

compiler-errors commented on compiler/rustc_trait_selection/src/solve/mod.rs:
```diff
@@ -235,17 +235,6 @@ impl<'a, 'tcx> EvalCtxt<'a, InferCtxt<'tcx>> {
 
 impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
-    #[instrument(level = "trace", skip(self, goals))]
-    fn add_goals(
```
I moved this with `add_goal`.

compiler-errors commented:
r? lcnr

lcnr commented on compiler/rustc_middle/src/ty/context.rs:
```diff
@@ -259,6 +259,49 @@ impl<'tcx> Interner for TyCtxt<'tcx> {
+    fn item_bounds(
+        self,
+        def_id: DefId,
+    ) -> ty::EarlyBinder<'tcx, impl Iterator<Item = ty::Clause<'tcx>>> {
```
why not `impl IntoIterator<Item = ..>`?

lcnr commented on compiler/rustc_trait_selection/src/solve/assembly/structural_traits.rs:
```diff
@@ -17,12 +16,15 @@ use crate::solve::EvalCtxt;
-    ty: Ty<'tcx>,
-) -> Result<Vec<ty::Binder<'tcx, Ty<'tcx>>>, NoSolution> {
+pub(in crate::solve) fn instantiate_constituent_tys_for_auto_trait<
+    Infcx: InferCtxtLike<Interner = I>,
+    I: Interner,
```
where bound :<

lcnr commented on compiler/rustc_trait_selection/src/solve/assembly/structural_traits.rs:
```diff
@@ -76,31 +78,36 @@ pub(in crate::solve) fn instantiate_constituent_tys_for_auto_trait<'tcx>(
-    match *ty.kind() {
+pub(in crate::solve) fn instantiate_constituent_tys_for_sized_trait<
+    Infcx: InferCtxtLike<Interner = I>,
+    I: Interner,
```
where-bound

lcnr commented on compiler/rustc_trait_selection/src/solve/assembly/structural_traits.rs:
```diff
@@ -160,11 +167,14 @@ pub(in crate::solve) fn instantiate_constituent_tys_for_sized_trait<'tcx>(
-    match *ty.kind() {
+pub(in crate::solve) fn instantiate_constituent_tys_for_copy_clone_trait<
+    Infcx: InferCtxtLike<Interner = I>,
+    I: Interner,
```
where-bound

lcnr commented on compiler/rustc_trait_selection/src/solve/assembly/structural_traits.rs:
```diff
@@ -643,17 +650,19 @@ fn coroutine_closure_to_ambiguous_coroutine<'tcx>(
-) -> Vec<Goal<'tcx, ty::Predicate<'tcx>>> {
+pub(in crate::solve) fn predicates_for_object_candidate<
+    Infcx: InferCtxtLike<Interner = I>,
+    I: Interner,
```
where-bound

lcnr commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1707,6 +1707,7 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                     ValuePairs::ExistentialProjection(_) => {
                         (false, Mismatch::Fixed("existential projection"))
                     }
+                    infer::DummyPair => (false, Mismatch::Fixed("values")),
```
why that instead of `ValuePairs::DummyPair`? and why not `bug!()`

compiler-errors replied:
oh lol i have no idea

lcnr commented on compiler/rustc_type_ir/src/inherent.rs:
```diff
@@ -198,16 +293,44 @@ pub trait Predicate<I: Interner<Predicate = Self>>:
 {
     fn is_coinductive(self, interner: I) -> bool;
+
+    fn allow_normalization(self) -> bool {
```
instead of having to add a FIXME: keep in sync, isn't it easier to only define that method in `rustc_middle`?

compiler-errors replied:
Right -- I guess I would like to eventually move this up into the solver so that r-a doesn't need to keep their own impl in sync, but we can do that later.

lcnr reviewed (commented):
some nits, then r=me

compiler-errors committed:
Address nits

- Remove the ValuePairs glob import
- Make DummyPairs -> ValuePairs::Dummy and make it bug more
- Fix WC
- Make interner return `impl IntoIterator`s

compiler-errors commented:
@bors r=lcnr

bors commented:
:pushpin: Commit c8e42065f0f1d1a5cdb29c4279b1f2c25bc8737b has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr c8e42065f0f1d1a5cdb29c4279b1f2c25bc8737b -->
<!-- homu: {"type":"Approved","sha":"c8e42065f0f1d1a5cdb29c4279b1f2c25bc8737b","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
